lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,259 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isLoopStatement(AstNode node) => node is DoStatement ||
+      node is ForEachStatement ||
+      node is ForStatement ||
+      node is WhileStatement;
+
+_needsLocalEnvironment(AstNode node) =>
+    node is ClassDeclaration ||
+    node is CompilationUnit ||
+    node is ConstructorDeclaration ||
+    node is FunctionDeclaration ||
+    node is MethodDeclaration;
```

Comment:
Just so it isn't forgotten, I'll repeat the request:

Please add documentation somewhere explaining when and/or why we need a local environment.